README
jxWeather v0.3.2.SNAPSHOT
Copyright (C) 2003, Azusa Pacific University and Daniel Berg (C) 2016

Original Contact:  David Castro <dcastro@apu.edu>
Contact : Daniel Berg g95232300@gmail.com

This is a fork of jweather (https://github.com/arimus/jweather). 

This fork mostly contains a simpler dependency graph with no oro or apache commons http client or any external jar. 
All features in the original project that had such dependencies are rewritten to be run with plain java 1.6.
Quite possibly we'll rewrite the whole thing and release a new one - but this is LGPL so here are my current modification published :)

jWeather is a Java library for parsing raw weather data (e.g. METAR, TAF).
Initially it will focus on parsing and providing an API for access to METAR
data, but may grow to include access to multiple reporting formats.

Please see LICENSE in this directory for jWeather licensing information.  

Thanks to the National Oceanic and Atmospheric Administration for the work they
have done to provide this data to the public.  More information regarding
weather data and the NOAA can be found at http://www.noaa.gov/

Information regarding METAR specification can be found at
http://www.ofcm.gov/fmh-1/fmh1.htm

Information regarding the TAF specification can be found at
http://metar.noaa.gov/table_master.jsp?sub_menu=yes&dir=./rel_docs/wsom/d31/&show=d31links.htm&title=title_related_docs

** NOTE **
This API absolutely will change until there is a stable release! Relying on it
to not change is probably NOT a safe bet.  It is an initial release, given as a
(hopefully) better than nothing option.  The plan is for this library to become
much cleaner and well-architected, but only time will tell.  The more you show
interest in this library/nudge me, the more likely that will be the case.
Comments and feedback greatly welcome.


Requirements:
  Java 2 SDK 1.6
  maven 3

Building:
  type 'mvn clean package'

  should have your jar file in ./target

Testing:
  optionally change the station id in the build.xml file.  To do this, change
  the argument specified for 'test' target to a valid station id (see
  stations.txt for a list of valid stations).

  Log4j setting must be provided by user

  with any luck, you should now be seeing parsed weather data

Notes:

  
Problems/questions/suggestions:
  daniel berg g95232300@gmail.com
 


Changelog
Changelog
---------
Changes in 0.3.2-SNAPSHOT
 * Removed dependencies on oro for regex, apache commons-httpclient, and log4j.
 
Changes since 0.2.4
	* added copyright/license text to all source files
	* updated license to LGPL from GPL
	* Replace URLConnection use with Apache commons-httpclient, which has
	*   support for connection timeout configuration
	* fixed a getVisibility() bug that was throwing an NPE

Changes since 0.2.3
	* Fixed CAVOK problem.  Was specifying 10 miles instead of 10 kilometers
	* Support for visibility tokens like '9999' added.  Similar to CAVOK, means
	* visibility greater than 10KM
	* Added support for NOSIG
	* Added support for NSC

Changes since 0.2.2
	* various bugfixes for international reporting stations
	* - CAVOK now supported (885981)
	* - visibility in kilometers and meters now supported (877179, 877181)
	* - wind speed in meters/second now supported (877183)
	* support for fractional visibilities (875903)
	* made thread safe (875919)

Changes since 0.2.1
	* Moved all METAR-related classes into the subpackage 'metar'
	* Split up METAR downloading and parsing
	* Added a file that contains all METAR-related constants
	* Added junit test file for the METAR classes (still needs work)
	* Added fix for fractional and "less than" visibilities
	* Miscellaneous bug fixes

	* special thanks goes to Dennis Keller <dennis@bullamanka.com> for
	  bug fixes and initiating a change in the structure of jweather that
	  should make things more maintainable for the future.

Changes since 0.2
	* Fixed a bug regarding variable wind direction.  "VRB" is listed only
	  in the case where the variable wind is <= 6 knots in speed.  Else, it
	  is coded in the token immediately following in the format
		F F F VF F F
		  n n n  x x x
	  Was only checking to see if the next token was variable wind
	  direction if VRB was set, which is incorrect.
	  
	  Thanks to bobzilla for identifying the problem.

Changes since 0.1
	* Removed all JDK 1.4 regexp calls (e.g. matches() & split()) and
	  replaced them with the equivalent Jakarta ORO calls
	* Bundling ORO with jweather



src/net/sf/jweather/Weather.java
/*
jWeather(TM) is a Java library for parsing raw weather data
Copyright (C) 2004 David Castro

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For more information, please email arimus@users.sourceforge.net
 */
package net.sf.jweather;

import java.io.*;
import java.net.*;
import java.util.logging.Logger;
import net.sf.jweather.metar.*;

/**
 * Responsible for downloading the METAR reports and feeding them to the Metar
 * class to be parsed
 *
 * <p>
 * <code>
 * Metar metar = Weather.getMetar("KLAX");<br>
 * metar.print();
 * </code>
 * </p>
 *
 * @author David Castro, dcastro@apu.edu
 * @version $Revision: 1.8 $
 * @see <a href="MetarFetcher.html">Metar</a>
 * @see <a href="Metar.html">Metar</a>
 */
public class Weather {
	private static Logger log = null;

	final static String httpMetarURL = "http://weather.noaa.gov/pub/data/observations/metar/stations/";
	// final static String httpMetarHostname = "weather.noaa.gov";
	// final static int httpMetarPort = 80;
	// final static String httpMetarPath =
	// "/pub/data/observations/metar/stations/";

	static {
		log = Logger.getLogger("net.sf.jweather");
		log.fine("Weather: instantiated");
	}

	public static Metar getMetar(String station) {
		return getMetar(station, 0);
	}

	public static Metar getMetar(String station, int timeout) {

		Metar metar = null;
		try {
			String metarData = MetarFetcher.fetch(station, timeout);
			if (metarData != null) {

				metar = MetarParser.parseRecord(metarData);

			}
		} catch (Exception e) {
			System.err.println("Weather: " + e);
			e.printStackTrace(System.err);
		}
		return metar;
	}
}


src/net/sf/jweather/metar/MetarFetcher.java
/*
jWeather(TM) is a Java library for parsing raw weather data
Copyright (C) 2004 David Castro

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For more information, please email arimus@users.sourceforge.net
*/
package net.sf.jweather.metar;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.logging.Logger;

/**
 * Responsible for downloading the METAR reports 
 *
 * <p>
 * <code>
 * String metarData = MetarFetcher.fetch("KLAX");<br>
 * </code>
 * </p>
 *
 * @author David Castro, dcastro@apu.edu
 * @version $Revision: 1.4 $
 * @see <a href="Metar.html">Metar</a>
 */
public class MetarFetcher {
	private static Logger log = null;
	private static String metarData = null;
	static final Charset defaultCharset = Charset.forName("UTF-8");
	final static String httpMetarURL = "http://weather.noaa.gov/pub/data/observations/metar/stations/";
	//final static String httpMetarHostname =  "weather.noaa.gov";
	//final static int    httpMetarPort     =  80;
	//final static String httpMetarPath     = "/pub/data/observations/metar/stations/";

	static {
    	log = Logger.getLogger("net.sf.jweather");
		log.fine("MetarFetcher: instantiated");
	}

	public static String fetch(String station)throws Exception {
		return fetch(station, 0);
	}

	public static String fetch(String station, int timeout) throws Exception{
		metarData = null;

		HttpURLConnection openConnection = null;
		try{
		URL url = new URL(httpMetarURL + station + ".TXT");
		 openConnection =(HttpURLConnection)url.openConnection();
		
			// set the timeout is specified
			if (timeout != 0) {
				log.fine("MetarFetch: setting timeout to '"+timeout+"' milliseconds");
				long start = System.currentTimeMillis();
				openConnection.setConnectTimeout(timeout);
				long end = System.currentTimeMillis();
				if (end - start < timeout) {
					openConnection.setReadTimeout((int)(end - start));
				} else {
					return null;
				}
			}
		

		// check that we didn't run out of retries
		if (openConnection.getResponseCode() != HttpURLConnection.HTTP_OK) {
			log.severe("failed to download station data for '"+station+"'");
			return null;
		} else {
			
			InputStream is = url.openStream();
			int l = openConnection.getContentLength();
			ByteArrayOutputStream bos = new ByteArrayOutputStream(l);
			int read = 0; byte[] b = new byte[l];
			while( -1 !=( read= is.read(b))){
				bos.write(b,0, read);
			}
			is.close();
			metarData = new String(bos.toByteArray(),getCharset( openConnection.getContentType())) + "\n";
			log.fine("MetarFetcher: metar data: " + metarData);
		}
		}finally{
			if(openConnection!=null)openConnection.disconnect();
		}
		return metarData;
	}
	static Charset getCharset(String contentType){
		
		String charset = "UTF-8";
		if(contentType==null || contentType.length()<2)
			return defaultCharset;
		String[] values = contentType.split(";");
		if(values.length>5)return defaultCharset;
		
		for (String value : values) {
		    value = value.trim();		    
		    if (value.toLowerCase().startsWith("charset=")) {
		        charset = value.substring("charset=".length());
		    }
		}

		if ("".equals(charset)) {
		    charset = "UTF-8"; //Assumption
		}
		return Charset.isSupported(charset)?Charset.forName(charset):defaultCharset;
		

	}
}


src/net/sf/jweather/metar/Metar.java
/*
jWeather(TM) is a Java library for parsing raw weather data
Copyright (C) 2004 David Castro

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For more information, please email arimus@users.sourceforge.net
*/
package net.sf.jweather.metar;

import java.util.Date;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.logging.Logger;


/**
 * Responsible for storing METAR data and providing methods for accessing
 * the data
 * @author David Castro, dcastro@apu.edu
 * @version $Revision: 1.12 $
 * @see <a href="Weather.html">Weather</a>
 * @see <a href="Obscuration.html">Obscuration</a>
 * @see <a href="RunwayVisualRange.html">RunwayVisualRange</a>
 * @see <a href="SkyCondition.html">SkyCondition</a>
 * @see <a href="WeatherCondition.html">WeatherCondition</a>
 */
public class Metar {
	Logger log;

  String reportString = null;
	String dateString = "";
	Date date = null;
	String reportModifier = "";
	String stationID = "";
	Integer windDirection = null;
	Integer windDirectionMin = null;
	Integer windDirectionMax = null;
	boolean windDirectionIsVariable = false;
	Float windSpeed = null; // (in knots x 1.1508 = MPH)
	Float windGusts = null; // (in knots x 1.1508 = MPH)
	boolean isCavok = false;
	Float visibilityMiles = null; // in miles
	Float visibilityKilometers = null; // in kilometers
	Float visibilityMeters = null; // in meters
	boolean visibilityLessThan = false;
	Float pressure = null;
	Float temperature = null;
	Float temperaturePrecise = null;
	Float dewPoint = null;
	Float dewPointPrecise = null;
	ArrayList weatherConditions = new ArrayList();
	ArrayList skyConditions = new ArrayList();
	ArrayList runwayVisualRanges = new ArrayList();
	ArrayList obscurations = new ArrayList();
	private boolean isNoSignificantChange = false;
  private String becoming = null;

	public Metar() {
    	log = Logger.getLogger("net.sf.jweather");
		log.fine("Metar: instantiated");
	}

    /**
     *
	 * @param value string that represents the date this METAR report was
	 * generated
     */
	protected void setDateString(String value) {
		this.dateString = value;
	}

    /**
     *
     * @return string that represents the date this METAR report was generated
     */
	public String getDateString() {
		return dateString;
	}

    /**
     *
	 * @param value the date this METAR report was generated
     */
	protected void setDate(Date value) {
		this.date = value;
	}

    /**
     *
     * @return the date this METAR report was generated
     */
	public Date getDate() {
		return date;
	}

    /**
     *
	 * @param value the modifier of the report, which specifies whether this
	 * report was an automated report or was a corrected report
     */
	protected void setReportModifier(String value) {
		this.reportModifier = value;
	}

    /**
     *
     * @return the modifier of the report, which specifies whether this report
     * was an automated report or was a corrected report
     */
	public String getReportModifier() {
		return reportModifier;
	}

    /**
     *
	 * @param value the station id of the station that generated this METAR
	 * report
     */
	protected void setStationID(String value) {
		this.stationID = value;
	}

    /**
     *
     * @return the station id of the station that generated this METAR report
     */
	public String getStationID() {
		return stationID;
	}

    /**
     *
     * @param value the direction the wind is blowing in (in degrees)
     */
	protected void setWindDirection(Integer value) {
		this.windDirection = value;
	}

    /**
     *
     * @return the direction the wind is blowing in (in degrees)
     */
	public Integer getWindDirection() {
		return windDirection;
	}

    /**
     *
	 * @param value the minimum wind direction (in degrees) for variable wind
	 * directions
     */
	protected void setWindDirectionMin(Integer value) {
		this.windDirectionMin = value;
	}

    /**
     *
	 * @return the minimum wind direction (in degrees) for variable wind
	 * directions
     */
	public Integer getWindDirectionMin() {
		return windDirectionMin;
	}

    /**
     *
	 * @param value the maximum wind direction (in degrees) for variable wind
	 * directions
     */
	protected void setWindDirectionMax(Integer value) {
		this.windDirectionMax = value;
	}

    /**
     *
	 * @return the maximum wind direction (in degrees) for variable wind
	 * directions
     */
	public Integer getWindDirectionMax() {
		return windDirectionMax;
	}

    /**
     *
     * @param value whether or not the wind direction is variable
     */
	protected void setWindDirectionIsVariable(boolean value) {
		this.windDirectionIsVariable = value;
	}

    /**
     *
     * @return whether or not the wind direction is variable
     */
	public boolean getWindDirectionIsVariable() {
		return windDirectionIsVariable;
	}

    /**
     *
     * @param value wind speed in knots
     */
	protected void setWindSpeed(Float value) {
		this.windSpeed = value;
	}

    /**
     *
     * @param value wind speed in meters per second
     */
	protected void setWindSpeedInMPS(Float value) {
		this.windSpeed = new Float(value.floatValue() / 0.5148);
	}

    /**
     *
     * @return wind speed in meters per second
     */
	public Float getWindSpeedInMPS() {
		return new Float(this.windSpeed.floatValue() * 0.5148);
	}

    /**
     *
     * @return wind speed in knots
     */
	public Float getWindSpeedInKnots() {
		return this.windSpeed;
	}

    /**
     *
     * @return wind speed in MPH
     */
	public Float getWindSpeedInMPH() {
		if (this.windSpeed == null) {
			return null;
		}

		double f = this.windSpeed.floatValue() * 1.1508;

		// round to the nearest MPH
		f = Math.round(f);

		return new Float(f);
	}

    /**
     *
     * @param value wind gust speed in knots
     */
	protected void setWindGusts(Float value) {
		this.windGusts = value;
	}

    /**
     *
     * @param value wind gust speed in meters per second
     */
	protected void setWindGustsInMPS(Float value) {
		this.windGusts = new Float(value.floatValue() / 0.5148);
	}

    /**
     *
     * @return wind gust speed in meters per second
     */
	public Float getWindGustsInMPS() {
		return new Float(this.windGusts.floatValue() * 0.5148);
	}

    /**
     *
     * @return wind gust speed in knots
     */
	public Float getWindGustsInKnots() {
		return this.windGusts;
	}

    /**
     *
     * @return wind gust speed in MPH
     */
	public Float getWindGustsInMPH() {
		if (this.windGusts == null) {
			return null;
		}

		double f = this.windGusts.floatValue() * 1.1508;

		// round to the nearest MPH
		f = Math.round(f);

		return new Float(f);
	}

    /**
	 * this function will also set visibility to 10KM, since
	 * CAVOK means visibility is greater than 10KM
     *
     * @param value boolean whether or not CAVOK was given
     */
	protected void setIsCavok(boolean value) {
		this.isCavok = value;
		// set visibility to 10
		setVisibilityInKilometers(new Float(10));
	}

    /**
     * @return value boolean whether or not CAVOK is true
     */
	public boolean getIsCavok() {
		return this.isCavok;
	}

    /**
     *
     * @param value visibility in miles
     */
	protected void setVisibility(Float value) {
		this.visibilityMiles = value;
		this.visibilityKilometers = null;
		this.visibilityMeters = null;
	}

    /**
     *
     * @param value visibility in kilometers
     */
	protected void setVisibilityInKilometers(Float value) {
		this.visibilityKilometers = value;
		this.visibilityMiles = null;
		this.visibilityMeters = null;
	}

    /**
     *
     * @param value visibility in meters
     */
	protected void setVisibilityInMeters(Float value) {
		this.visibilityMeters = value;
		this.visibilityMiles = null;
		this.visibilityKilometers = null;
	}

    /**
     *
     * @return visibility in miles
     */
	public Float getVisibility() {
		if (visibilityMiles != null) {
			return visibilityMiles;
		} else if (visibilityKilometers != null) {
			return new Float(visibilityKilometers.floatValue() / 1.609344);
		} else if (visibilityMeters != null) {
			return new Float(visibilityMeters.floatValue() / 1609.344);
		}
		return null;
	}

    /**
     *
     * @return visibility in kilometers
     */
	public Float getVisibilityInKilometers() {
		if (visibilityKilometers != null) {
			return visibilityKilometers;
		} else if (visibilityMeters != null) {
			return new Float(visibilityMeters.floatValue() / 1000);
		} else if (visibilityMiles != null) {
			return new Float(visibilityMiles.floatValue() * 1.609344);
		}
		return null;
	}

    /**
     *
     * @return visibility in meters
     */
	public Float getVisibilityInMeters() {
		if (visibilityMeters != null) {
			return visibilityMeters;
		} else if (visibilityKilometers != null) {
			return new Float(visibilityKilometers.floatValue() * 1000);
		} else if (visibilityKilometers != null) {
			return new Float(visibilityKilometers.floatValue() * 1609.344);
		}
		return null;
	}

    /**
     *
     * @param value visibility less than
     */
	protected void setVisibilityLessThan(boolean value) {
		this.visibilityLessThan = value;
	}

    /**
     *
     * @return visibility less than
     */
	public boolean getVisibilityLessThan() {
		return visibilityLessThan;
	}

    /**
     *
     * @param value pressure in inches Hg
     */
	protected void setPressure(Float value) {
	  // Round so we're consistent
          if( value != null )
	  {
            float val = value.floatValue();
            val = (float)Math.round(val*100)/100;
            this.pressure = new Float(val);
	  }
	  else {
	    this.pressure = value;
	  }
	}

    /**
     *
     * @return pressure in inches Hg
     */
	public Float getPressure() {
		return pressure;
	}

    /**
     *
     * @return pressure in hPa
     */
    public Integer getPressureInHectoPascals() {
	    if( pressure == null ) return null;

      // convert to hPa
      float val = pressure.floatValue()*33.8639F;  
      // round
      return new Integer(Math.round(val)); 
    }

    /**
     *
     * @param value temperature in celsius
     */
	protected void setTemperature(Float value) {
		this.temperature = value;
	}

    /**
     *
     * @return temperature in celsius
     */
	public Float getTemperatureInCelsius() {
		return this.temperature;
	}

    /**
     *
     * @return temperature in fahrenheit
     */
	public Float getTemperatureInFahrenheit() {
		if (this.temperature == null) {
			return null;
		}

		// round to the nearest 1/10th
		float f = (float)Math.round((this.temperature.floatValue()*9/5+32)*10)/10;

		return new Float(f);
	}

    /**
     *
     * @param value precise temperature in celsius
     */
	protected void setTemperaturePrecise(Float value) {
		this.temperaturePrecise = value;
	}

    /**
     *
     * @return precise temperature in celsius (nearest 1/10th degree)
     */
	public Float getTemperaturePreciseInCelsius() {
		return this.temperaturePrecise;
	}

    /**
     *
     * @return precise temperature in fahrenheit (nearest 1/10th degree)
     */
	public Float getTemperaturePreciseInFahrenheit() {
		if (this.temperaturePrecise == null) {
			return null;
		}

		// round to the nearest 1/10th
		float f = (float)Math.round((this.temperaturePrecise.floatValue()*9/5+32)*10)/10;
		return new Float(f);
	}

    /**
     *
     * @return most precise temperature in celsius (nearest 1/10th degree)
     */
	public Float getTemperatureMostPreciseInCelsius() {
		if (this.temperaturePrecise != null) {
			return this.temperaturePrecise;
		} else {
			return this.temperature;
		}
	}

    /**
     *
     * @return most precise temperature in fahrenheit (nearest 1/10th degree)
     */
	public Float getTemperatureMostPreciseInFahrenheit() {
		if (this.temperaturePrecise != null) {
			// round to the nearest 1/10th
			float f = (float)Math.round((this.temperaturePrecise.floatValue()*9/5+32)*10)/10;

			return new Float(f);
		} else if (this.temperature != null) {
			// round to the nearest 1/10th
			float f = (float)Math.round((this.temperature.floatValue()*9/5+32)*10)/10;

			return new Float(f);
		} else {
			return null;
		}
	}

    /**
     *
     * @param value dew point in celsius
     */
	protected void setDewPoint(Float value) {
		this.dewPoint = value;
	}

    /**
     *
     * @return dew point in celsius
     */
	public Float getDewPointInCelsius() {
		return this.dewPoint;
	}

    /**
     *
     * @return dew point in fahrenheit
     */
	public Float getDewPointInFahrenheit() {
		if (this.dewPoint == null) {
			return null;
		}

		// round to the nearest 1/10th
		float f = (float)Math.round((this.dewPoint.floatValue()*9/5+32)*10)/10;

		return new Float(f);
	}

    /**
     *
     * @param value precise dew point in celsius
     */
	protected void setDewPointPrecise(Float value) {
		this.dewPointPrecise = value;
	}

    /**
     *
     * @return dew point in celsius (nearest 1/10th degree)
     */
	public Float getDewPointPreciseInCelsius() {
		return this.dewPointPrecise;
	}

    /**
     *
     * @return dew point in fahrenheit (nearest 1/10th degree)
     */
	public Float getDewPointPreciseInFahrenheit() {
		if (this.dewPointPrecise == null) {
			return null;
		}

		// round to the nearest 1/10th
		float f = (float)Math.round((this.dewPointPrecise.floatValue()*9/5+32)*10)/10;

		return new Float(f);
	}

    /**
     *
     * @return most precise dew point in celsius (nearest 1/10th degree)
     */
	public Float getDewPointMostPreciseInCelsius() {
		if (this.dewPointPrecise != null) {
			return this.dewPointPrecise;
		} else {
			return this.dewPoint;
		}
	}

    /**
     *
     * @return most precise dew point in fahrenheit (nearest 1/10th degree)
     */
	public Float getDewPointMostPreciseInFahrenheit() {
		if (this.dewPointPrecise != null) {
			// round to the nearest 1/10th
			float f = (float)Math.round((this.dewPointPrecise.floatValue()*9/5+32)*10)/10;

			return new Float(f);
		} else if (this.dewPoint != null) {
			// round to the nearest 1/10th
			float f = (float)Math.round((this.dewPoint.floatValue()*9/5+32)*10)/10;

			return new Float(f);
		} else {
			return null;
		}
	}

    /**
     *
     * @param value whether or not the weather has changed significantly
     */
	protected void setIsNoSignificantChange(boolean value) {
		this.isNoSignificantChange = value;
	}

  /**
   * 
   * @return whether or not there has been a significant change in
	 * weather
   */
	public boolean getIsNoSignificantChange() {
		return isNoSignificantChange;
	}

  /**
   * Get the "BECMG" string in raw form. We don't have the ability to parse
   * this string yet, so it's being stored in raw form for now.
   * @return Returns The becoming (BECMG) String in raw form.
   */
  public String getBecoming() {
    return becoming;
  }
  /**
   * Set the "BECMG" string in raw form. We don't have the ability to parse
   * this string yet, so it's being stored in raw form for now.
   * @param becoming The becoming (BECMG) String in raw form.
   */
  public void setBecoming(String becoming) {
    this.becoming = becoming;
  }
	/**
	 *
	 * @param wc a WeatherCondition object
     * @see WeatherCondition
	 */
	public void addWeatherCondition(WeatherCondition wc) {
		weatherConditions.add(wc);
	}

	/**
	 *
	 * @return a WeatherCondition object
     * @see WeatherCondition
	 */
	public WeatherCondition getWeatherCondition(int i) {
		if (weatherConditions.size() >= i) {
			return (WeatherCondition)weatherConditions.get(i);
		} else {
			return null;
		}
	}

    /**
     *
     * @return an ArrayList of WeatherCondition objects
     * @see WeatherCondition
     */
	public ArrayList getWeatherConditions() {
		return weatherConditions;
	}

	/**
	 *
	 * @param sc a SkyCondition object
     * @see SkyCondition
	 */
	public void addSkyCondition(SkyCondition sc) {
		skyConditions.add(sc);
	}

	/**
	 *
	 * @return a SkyCondition object
     * @see SkyCondition
	 */
	public SkyCondition getSkyCondition(int i) {
		if (skyConditions.size() >= i) {
			return (SkyCondition)skyConditions.get(i);
		} else {
			return null;
		}
	}

    /**
     *
     * @return an ArrayList of SkyCondition objects
     * @see SkyCondition
     */
	public ArrayList getSkyConditions() {
		return skyConditions;
	}

	public int getMinimumCloudCoverSCR(){
		int minimumHeight = Integer.MAX_VALUE;
		Iterator it = skyConditions.iterator();
		while(it.hasNext()){
			SkyCondition sc = (SkyCondition)it.next();
			if( sc.isClear()|| sc.isNoSignificantClouds() || sc.isFewClouds() )
				continue;
			if(sc.getHeight()<minimumHeight)minimumHeight=sc.getHeight();
		}
		
		return(minimumHeight==Integer.MAX_VALUE)?-1:minimumHeight;
		
	}
	/**
	 *
	 * @param rvr a RunwayVisualRange object
     * @see RunwayVisualRange
	 */
	public void addRunwayVisualRange(RunwayVisualRange rvr) {
		runwayVisualRanges.add(rvr);
	}

	/**
	 *
	 * @return a RunwayVisualRange object
     * @see RunwayVisualRange
	 */
	public RunwayVisualRange getRunwayVisualRange(int i) {
		if (runwayVisualRanges.size() >= i) {
			return (RunwayVisualRange) runwayVisualRanges.get(i);
		} else {
			return null;
		}
	}

    /**
     *
     * @return an ArrayList of RunwayVisualRange objects
     * @see RunwayVisualRange
     */
	public ArrayList getRunwayVisualRanges() {
		return runwayVisualRanges;
	}

	/**
	 *
	 * @param o an Obscuration object
     * @see Obscuration
	 */
	public void addObscuration(Obscuration o) {
		obscurations.add(o);
	}

	/**
	 *
	 * @return a Obscuration object
     * @see Obscuration
	 */
	public Obscuration getObscuration(int i) {
		if (obscurations.size() >= i) {
			return (Obscuration)obscurations.get(i);
		} else {
			return null;
		}
	}

    /**
     *
     * @return an ArrayList of Obscuration objects
     * @see Obscuration
     */
	public ArrayList getObscurations() {
		return obscurations;
	}
	
    /**
     * display metar data in a human-readable format
     */
    public void print() {
		System.out.println("station id : " + getStationID());
		System.out.println("wind dir   : " + getWindDirection() + " degrees");
		System.out.println("wind speed : " + getWindSpeedInMPH() + " mph, " +
				                             getWindSpeedInKnots() + " knots");
		System.out.println("wind gusts : " + getWindGustsInMPH() + " mph, " +
				                             getWindGustsInKnots() + " knots");
		if (!getVisibilityLessThan()) {
			System.out.println("visibility : " + getVisibility() + " mile(s)");
		} else {
			System.out.println("visibility : < " + getVisibility() + " mile(s)");
		}

		System.out.println("pressure   : " + getPressure() + " in Hg");
		System.out.println("temperaturePrecise: " +
				           getTemperaturePreciseInCelsius() + " C, " +
				           getTemperaturePreciseInFahrenheit() + " F");
		System.out.println("temperature: " +
				           getTemperatureInCelsius() + " C, " +
				           getTemperatureInFahrenheit() + " F");
		System.out.println("temperatureMostPrecise: " +
				           getTemperatureMostPreciseInCelsius() + " C, " +
				           getTemperatureMostPreciseInFahrenheit() + " F");

		System.out.println("dewPointPrecise: " +
				           getDewPointPreciseInCelsius() + " C, " +
				           getDewPointPreciseInFahrenheit() + " F");
		System.out.println("dewPoint: " +
				           getDewPointInCelsius() + " C, " +
				           getDewPointInFahrenheit() + " F");
		System.out.println("dewPointMostPrecise: " +
				           getDewPointMostPreciseInCelsius() + " C, " +
				           getDewPointMostPreciseInFahrenheit() + " F");

		if (getWeatherConditions() != null) {
			Iterator i = getWeatherConditions().iterator();
			while (i.hasNext()) {
				WeatherCondition weatherCondition = (WeatherCondition)i.next();
				System.out.println(weatherCondition.getNaturalLanguageString());
			}
		}
		if (getSkyConditions() != null) {
			Iterator i = getSkyConditions().iterator();
			while (i.hasNext()) {
				SkyCondition skyCondition = (SkyCondition)i.next();
				System.out.println(skyCondition.getNaturalLanguageString());
			}
		}
	}
  
  /**
   * Retrieve the full report string value. This is the report in its
   * original form
   * 
   * @return The original report string.
   */
  public String getReportString() {
    return reportString;
  }

  /**
   * Set the full report string value. This is the report in its
   * original form
   * 
   * @param string  The original report string.
   */
  public void setReportString(String string) {
    reportString = string;
  }

}

